pkg/cleaning/git_history_based_cleanup/repo_with_cache.go

```diff
@@ -0,0 +1,67 @@
+package git_history_based_cleanup
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+)
+
+type GitRepo interface {
+	PlainOpen() (*git.Repository, error)
+}
+
+type GitRepositoryWithCache struct {
+	GitRepo     *git.Repository
+	CommitCache sync.Map
+	mu          sync.Mutex
+}
+
+func NewGitRepositoryWithCache(gitRepo GitRepo) (*GitRepositoryWithCache, error) {
+	gitRepository, err := gitRepo.PlainOpen()
+	if err != nil {
+		return nil, fmt.Errorf("git plain open failed: %w", err)
+	}
+	return &GitRepositoryWithCache{
+		GitRepo: gitRepository,
+	}, nil
+}
+
+// CommitObject tries to load a commit object from the cache.
+// If the commit is not found in the cache, it locks the critical section where we try to load commit data from disk.
+// It doesn't actually protect any data and works like a semaphore.
+// Locking particular commit only is not safe because of inner go-git logic.
+// We need this workaround because go-git is not thread-safe. It should be fixed in go-git v6.
+// Ref: https://github.com/go-git/go-git/issues/773
+func (g *GitRepositoryWithCache) CommitObject(commitHash plumbing.Hash) (*object.Commit, error) {
+	if c, ok := g.getFromCache(commitHash); ok {
+		return c, nil
+	}
+	g.mu.Lock()
+	defer g.mu.Unlock()
```

Comment:
```suggestion
	if c, ok := g.getFromCache(commitHash); ok {
		return c, nil
	}
	
	g.mu.Lock()
	defer g.mu.Unlock()
	
	if c, ok := g.getFromCache(commitHash); ok {
		return c, nil
	}
```

Reply:
Double-check the cache after acquiring the lock to avoid redundant Git operations.